modules/appeals_api/app/controllers/concerns/appeals_api/pdf_downloads.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+require 'common/file_helpers'
+require 'pdf_fill/filler'
+
+# rubocop:disable Metrics/ModuleLength
+module AppealsApi
+  module PdfDownloads
+    extend ActiveSupport::Concern
+
+    included do
+      def render_appeal_pdf_download(appeal, filename)
+        return render_pdf_download_expired(appeal) if expired?(appeal)
+
+        return render_pdf_download_unauthorized unless download_authorized?(appeal)
+
+        return render_pdf_download_not_ready(appeal) unless submitted?(appeal)
+
+        pdf_path = PdfDownloads.watermark(
+          AppealsApi::PdfConstruction::Generator.new(appeal, pdf_version: appeal.pdf_version).generate,
+          filename
+        )
+
+        send_file(pdf_path, type: 'application/pdf; charset=utf-8', filename:)
+      end
+    end
+
+    # Creates a copy of the input PDF with a watermark on each page
+    def self.watermark(input_path, output_path = "#{Common::FileHelpers.random_file_path}.pdf")
+      num_pages = PDF::Reader.new(input_path).pages.length
+      pdftk = PdfFill::Filler::PDF_FORMS
+      stamp_path = "#{Common::FileHelpers.random_file_path}-watermark.pdf"
+
+      Prawn::Document.generate(stamp_path, margin: [30, 30]) do |pdf|
+        num_pages.times do
+          pdf.image WATERMARK_PATH, fit: [pdf.bounds.width, pdf.bounds.height], position: :center, vposition: :center
+          pdf.start_new_page # Final extra page won't be added to output
+        end
+      end
+
+      pdftk.multistamp(input_path, stamp_path, output_path)
+      FileUtils.rm_f(stamp_path)
+
+      output_path
+    end
+
+    UNSUBMITTED_STATUSES = %w[pending submitting error].freeze
+    WATERMARK_PATH = AppealsApi::Engine.root.join('config', 'images', 'va_seal.png')
+
+    def download_authorized?(appeal)
```

Comment:
Curious to hear feedback on this method, which determines whether the `X-VA-ICN` header provided by the user authorizes them to download the PDF:
- Is 401 the code to return when this returns false? (it's what I have right now)
- Is the logic here right? I'm checking the PII for a match with the user-supplied X-VA-ICN header in this order:
    - The `veteran_icn` value stored on the appeal by our ICN adding worker, if set
    - The `X-VA-ICN` header stored on the appeal in its auth_headers, if available
    - The `X-VA-SSN` header stored on the appeal in its auth_headers (compared with X-VA-ICN throuh an MPI lookup), if available
    - The `X-VA-File-Number` header stored on the appeal in its auth_headers, if available
        - This was the check I was most unsure about - is it right to compare this to the SSN from the MPI profile like I do here?

Reply:
(sneaky admin edit to number the questions you had for easier reference)

For `2.iv.a`: FileNumber is usually, but not always, the SSN - so it is not correct to validate the `X-VA-FileNumber` header against the SSN from MPI. MPI's data packet may have FileNumber as a field, however, which could be used but feels redundant with the fact that we know we will always receive the X-VA-ICN header/param

Reply:
@cilestin for NODs created via decision reviews v2, it seems like we may not always have an ICN or an SSN - the only required headers for that endpoint are File Number, First Name, and Last name. When you say that we will always receive the ICN, do you mean that the ICN will be reliably available on the MPI profile for that appeal, even if they didn't initially supply X-VA-ICN when creating the appeal?

Regardless, I'll remove the check comparing SSN and file number - I also haven't yet found an example of an MPI profile with file number in it, but I'll look around a bit.

Reply:
Ah, right. I always forget NOD uses FileNumber and not SSN. We should probably check w/ Alejandro to see whether we can get away with ICN -> SSN checks, or if there's still compelling reason to keep using FileNumber other than that it's what the PDF form expects. If they still require it, we may well need to require both ICN and FileNumber from the consumer for NODs :(